File: runtimes/jvm/src/com/flint/rt/Roots.java
package com.flint.rt;

import java.util.ArrayList;

/// Everything the collector must find, ported from `runtime/src/gc.rs`.
///
/// ## The decision everything else rests on
///
/// `doc/decisions/0001`: values live in explicit arrays the collector walks,
/// never in host locals. On wasm that was forced -- wasm locals are not
/// scannable. Here it is a CHOICE, and the same one, because it is what makes
/// the collector identical across the three runtimes rather than merely
/// equivalent.
///
/// * the VALUE STACK is the interpreter's operands and locals;
/// * the SHADOW STACK is what Java code roots across an allocation, the
///   analogue of the Rust's `push`/`r`/`pop_to`;
/// * globals, consts and singletons are the image's.
public final class Roots {
    /// The interpreter's operand stack. `stackTop` is the live prefix; nothing
    /// above it is scanned, which is what lets a frame drop its operands by
    /// moving one integer.
    public long[] stack = new long[1024];
    public int stackTop = 0;

    /// Values held across an allocation by Java code. A `Value` in a Java local
    /// does not survive a collection -- the collector cannot see it and cannot
    /// update it -- so anything live across an `alloc` goes here.
    public long[] shadow = new long[256];
    public int shadowTop = 0;

    /// The intern tables. WEAK, and scanned by the collector rather than
    /// traced: an entry whose value died is dropped, which is what lets every
    /// short string and keyword be interned without the table being a leak.
    public final Interns[] interns = Interns.tables();

    public long[] globals = new long[0];
    public long[] consts = new long[0];
    /// Sized AND FILLED WITH NIL at construction.
    ///
    /// Both halves matter. Sized, because `SING_BINDINGS` is read before any
    /// image has been loaded and an empty array would be an index error rather
    /// than an absent binding. Filled, because a zero-filled array of values is
    /// NOT a nil-filled one: 0 is the bit pattern of `+0.0`, so an unset slot
    /// read back as the DOUBLE ZERO and `dyn-bindings` answered a number where
    /// a map was expected. The failure surfaced three frames away as "assoc
    /// onto a double", which is why the slot is written rather than left.
    public long[] singletons = newSingletons();

    static long[] newSingletons() {
        long[] s = new long[Rt.SING_COUNT];
        java.util.Arrays.fill(s, Val.NIL);
        return s;
    }

    /// Old objects holding a young pointer. The generational invariant: an old
    /// object pointing at a young one MUST be in here, or the young one is
    /// never traced, dies, and leaves a stale pointer in something still live.
    public final ArrayList<Long> remembered = new ArrayList<>();

    public void vpush(long v) {
        if (stackTop == stack.length) {
            long[] bigger = new long[stack.length * 2];
            System.arraycopy(stack, 0, bigger, 0, stackTop);
            stack = bigger;
        }
        stack[stackTop++] = v;
    }

    public long vpop() { return stack[--stackTop]; }

    /// Root `v` and return its index. Read it back with `r(i)` AFTER any
    /// allocation -- that is the whole discipline, and reading the Java local
    /// instead is the bug this exists to prevent.
    public int push(long v) {
        if (shadowTop == shadow.length) {
            long[] bigger = new long[shadow.length * 2];
            System.arraycopy(shadow, 0, bigger, 0, shadowTop);
            shadow = bigger;
        }
        shadow[shadowTop] = v;
        return shadowTop++;
    }

    public long r(int i) { return shadow[i]; }
    public void setR(int i, long v) { shadow[i] = v; }
    public int mark() { return shadowTop; }
    public void popTo(int n) { shadowTop = n; }

    /// Every root, for the collector to read and REWRITE.
    ///
    /// One loop in one place. Skipping an array would not fail here -- it would
    /// collect those objects out from under the program and fail somewhere
    /// else, later, as a corrupted value in code that did nothing wrong.
    interface Visitor { long visit(long v); }

    void forEach(Visitor f) {
        for (int i = 0; i < stackTop; i++) stack[i] = f.visit(stack[i]);
        for (int i = 0; i < shadowTop; i++) shadow[i] = f.visit(shadow[i]);
        for (int i = 0; i < globals.length; i++) globals[i] = f.visit(globals[i]);
        for (int i = 0; i < consts.length; i++) consts[i] = f.visit(consts[i]);
        for (int i = 0; i < singletons.length; i++) singletons[i] = f.visit(singletons[i]);
    }
}

File: runtimes/clr/src/rt/Roots.cs
using System.Collections.Generic;

namespace Flint.Rt;

/// Everything the collector must find, ported from `runtime/src/gc.rs`.
///
/// `DECISIONS.md#dispatch`: values live in explicit arrays the collector walks,
/// never in host locals. On wasm that was forced -- wasm locals are not
/// scannable. Here it is a CHOICE, and the same one, because it is what makes
/// the collector identical across the three runtimes rather than merely
/// equivalent.
public sealed class Roots {
    /// The interpreter's operand stack. `StackTop` is the live prefix; nothing
    /// above it is scanned.
    public long[] Stack = new long[1024];
    public int StackTop;

    /// Values held across an allocation by host code. A value in a CLR local
    /// does not survive a collection -- the collector cannot see it and cannot
    /// update it -- so anything live across an `Alloc` goes here.
    public long[] Shadow = new long[256];
    public int ShadowTop;

    /// What this executor shares with every other in the same sandbox: var
    /// slots, constants, singletons and the intern tables. One object, pointed
    /// at by all of them.
    public Shared shared = new Shared();

    /// The runtime these roots belong to, so that registering a new executor
    /// can flip every peer's `safepoints` flag in one place.
    public Rt owner;

    /// Old objects holding a young pointer. An old object pointing at a young
    /// one MUST be in here, or the young one is never traced, dies, and leaves
    /// a stale pointer in something still live.
    public readonly List<long> Remembered = new();

    public void VPush(long v) {
        if (StackTop == Stack.Length) System.Array.Resize(ref Stack, Stack.Length * 2);
        Stack[StackTop++] = v;
    }

    public long VPop() => Stack[--StackTop];

    /// Root `v` and return its index. Read it back with `R(i)` AFTER any
    /// allocation -- that is the whole discipline, and reading the host local
    /// instead is the bug this exists to prevent.
    public int Push(long v) {
        if (ShadowTop == Shadow.Length) System.Array.Resize(ref Shadow, Shadow.Length * 2);
        Shadow[ShadowTop] = v;
        return ShadowTop++;
    }

    public long R(int i) => Shadow[i];
    public void SetR(int i, long v) => Shadow[i] = v;
    public int Mark() => ShadowTop;
    public void PopTo(int n) => ShadowTop = n;

    /// Every root, for the collector to read and REWRITE. One loop in one
    /// place: skipping an array would collect those objects out from under the
    /// program and fail somewhere else, later.
    public delegate long Visitor(long v);

    public void ForEach(Visitor f) {
        // This executor's own.
        for (int i = 0; i < StackTop; i++) Stack[i] = f(Stack[i]);
        for (int i = 0; i < ShadowTop; i++) Shadow[i] = f(Shadow[i]);
        // The sandbox's.
        long[] g = shared.Globals, c = shared.Consts, sg = shared.Singletons;
        for (int i = 0; i < g.Length; i++) g[i] = f(g[i]);
        for (int i = 0; i < c.Length; i++) c[i] = f(c[i]);
        for (int i = 0; i < sg.Length; i++) sg[i] = f(sg[i]);
        // Every OTHER executor in this sandbox (`DECISIONS.md#drivers`). A
        // collection happens with all of them PARKED at a safepoint, so nothing
        // is mutating these while they are walked.
        //
        // Skipping one would not fail here. It would collect that thread's live
        // objects out from under it and fail somewhere else, later, as a
        // corrupted value in code that did nothing wrong.
        foreach (Roots e in shared.others) {
            if (ReferenceEquals(e, this)) continue;
            // A parked executor's roots do not change. If they have, the
            // collector is walking a thread that is still RUNNING -- say so
            // here rather than as an index error four frames down.
            if (e.StackTop > e.Stack.Length) {
                throw new System.InvalidOperationException(
                    "flint: scanning a RUNNING executor: StackTop " + e.StackTop
                    + " past len " + e.Stack.Length);
            }
            for (int i = 0; i < e.StackTop; i++) e.Stack[i] = f(e.Stack[i]);
            for (int i = 0; i < e.ShadowTop; i++) e.Shadow[i] = f(e.Shadow[i]);
        }
    }

    /// Every executor's remembered set, drained together.
    ///
    /// Only correct during a collection, which is the only time every other
    /// executor is stopped. Draining one and not the rest would LOSE
    /// old-to-young edges another thread recorded, and a lost edge is a young
    /// object collected while an old one still points at it.
    public List<long> DrainRemembered() {
        var outv = new List<long>(Remembered);
        Remembered.Clear();
        foreach (Roots e in shared.others) {
            if (ReferenceEquals(e, this)) continue;
            outv.AddRange(e.Remembered);
            e.Remembered.Clear();
        }
        return outv;
    }
}
